pkg/services/govmomi/vcenter/clone.go

```diff
@@ -390,6 +408,117 @@ func getDiskConfigSpec(disk *types.VirtualDisk, diskCloneCapacityKB int64) (type
+					DiskMode:        string(types.VirtualDiskModePersistent),
+					ThinProvisioned: types.NewBool(true),
+					VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
+						FileName: "",
+					},
+				},
+				ControllerKey: controller.GetVirtualController().Key,
+			},
+			CapacityInKB: int64(dataDisk.SizeGiB) * 1024 * 1024,
+		}
+
+		// Assign unit number to the next slot on the controller.
+		if err := assignUnitNumber(ctx, dev, devices, additionalDisks, controller, unit); err != nil {
+			log.Error(err, "unable to assign unit number")
+			return nil, err
+		}
+		unit = *dev.UnitNumber
+
+		diskConfigSpec := types.VirtualDeviceConfigSpec{
+			Device:        dev,
+			Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+			FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+		}
+
+		log.Info("Generated device", "dev", dev)
+
+		additionalDisks = append(additionalDisks, &diskConfigSpec)
+	}
+
+	return additionalDisks, nil
+}
+
+// assignUnitNumber assigns a controller unit number to a device.
+func assignUnitNumber(ctx context.Context, device types.BaseVirtualDevice, existingDevices object.VirtualDeviceList, newDevices []types.BaseVirtualDeviceConfigSpec, controller types.BaseVirtualController, offset int32) error {
+	if offset > maxUnitNumber {
+		return errors.Errorf("unable to assign unit number due to offset %v exceeding max allowed of %v", offset, maxUnitNumber)
+	}
+
+	if controller == nil {
+		return errors.Errorf("unable to assign unit number due to controller parameter being nil")
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	vd := device.GetVirtualDevice()
+	vd.ControllerKey = controller.GetVirtualController().Key
+	vd.UnitNumber = &offset
+
+	units := make([]bool, maxUnitNumber)
+
+	for i := 0; i < int(offset); i++ {
+		units[i] = true
+	}
+
+	sc, ok := controller.(types.BaseVirtualSCSIController)
+	if ok {
+		//  The SCSI controller sits on its own bus
+		log.V(4).Info(fmt.Sprintf("Marking SCSI Controller's unit number: %d", sc.GetVirtualSCSIController().ScsiCtlrUnitNumber))
+		units[sc.GetVirtualSCSIController().ScsiCtlrUnitNumber] = true
+	}
+
+	key := controller.GetVirtualController().Key
+
+	// Check all existing devices
+	for _, device := range existingDevices {
+		d := device.GetVirtualDevice()
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	// Check new devices
+	for _, device := range newDevices {
+		d := device.GetVirtualDeviceConfigSpec().Device.GetVirtualDevice()
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	// Assign first unused unit number
+	for unit, used := range units {
+		if !used {
+			unit32 := int32(unit)
+			vd.UnitNumber = &unit32
+			log.V(4).Info(fmt.Sprintf("Determined next available unit number: %d", unit32))
+			return nil
+		}
+	}
+
+	// If we are here, we did not find a unit number.  Return error.
+	return errors.Errorf("unable to find available unit number")
```

Comment:
```suggestion
	return errors.New("all unit numbers are already in-use")
```

Not sure if that's a good error message though.

Reply:
Yeah, I would almost think the template image being already full before attempting to add an additional should never happen, but this is a good safety net in case something strange happens.  I was almost leaning towards `no unit numbers are available for use", but if your suggestion makes sense to you, I am ok with it.  I'll change this to the new text unless someone else has a better suggestion.